packages/nitro-protocol/contracts/ninja-nitro/SingleChannelAdjudicator.sol

```diff
@@ -0,0 +1,476 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.4;
+pragma experimental ABIEncoderV2;
+
+import '../interfaces/IAssetHolder.sol';
+import '../ForceMove.sol';
+import '../Outcome.sol';
+import './AdjudicatorFactory.sol';
+import './OutcomeTransformations.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+contract SingleChannelAdjudicator is ForceMove, OutcomeTransformations {
+    address public immutable adjudicatorFactoryAddress;
+
+    constructor(address a) {
+        adjudicatorFactoryAddress = a;
+    }
+
+    receive() external payable {} // solhint-disable-line no-empty-blocks
+
+    struct ChannelDataAlt {
+        uint48 turnNumRecord;
+        uint48 finalizesAt;
+        bytes32 stateHash; // keccak256(abi.encode(State))
+        address challengerAddress;
+        bytes outcomeBytes;
+    }
+
+    /**
+     * @notice Verifies a conclusion proof, pays out all assets and selfdestructs
+     * @dev Verifies a conclusion proof, pays out all assets and selfdestructs
+     * @param largestTurnNum The largest turn number of the submitted states; will overwrite the stored value of `turnNumRecord`.
+     * @param fixedPart Data describing properties of the state channel that do not change with state updates.
+     * @param appPartHash The keccak256 of the abi.encode of `(challengeDuration, appDefinition, appData)`. Applies to all states in the finalization proof.
+     * @param outcomeBytes abi.encode of an array of Outcome.OutcomeItem structs.
+     * @param numStates The number of states in the finalization proof.
+     * @param whoSignedWhat An array denoting which participant has signed which state: `participant[i]` signed the state with index `whoSignedWhat[i]`.
+     * @param sigs An array of signatures that support the state with the `largestTurnNum`.
+     */
+    function concludeAndTransferAll(
```

Comment:
It seems to me like this should be `concludeAndTransfer`, and accept an array of indices for each asset outcome. (An empty indices object can indicate a `TransferAll` operation.)

This simplifies the exit algorithm:
1. Call `concludeAndTransfer`, transferring out as many outcome items as they're willing.
2. While the adjudicator exists, call `transfer` to iteratively transfer out remaining funds.

Currently,
- one needs to differentiate between two cases: "I am transferring all" vs "I am not transferring all"
- in the second case, one needs an extra transaction to conclude the channel, wasting a 21k base fee

Reply:
> It seems to me like this should be concludeAndTransfer, and accept an array of indices for each asset outcome. (An empty indices object can indicate a TransferAll operation.)

Agreed. 

> This simplifies the exit algorithm:

> Call concludeAndTransfer, transferring out as many outcome items as they're willing.
While the adjudicator exists, call transfer to iteratively transfer out remaining funds.

I'm not sure I follow this?

Reply:
Currently, the wallet needs to decide between:
- calling `concludePushOutcomeAndTransferAll`
- 1. calling `conclude`; 2. calling `pushOutcome`; 3. calling `transfer` iteratively until the channel is fully de-funded

With the proposed change, the wallet does not need to decide between these two strategies. The wallet can `concludeAndTransfer`, passing indices for which outcome items should be tackled. If the channel is not fully-defunded, the wallet continues to call `transfer` with more indices.

Reply:
I think what is confusing me is that in both of your strategies, you mention one wallet making calls to fully defund a channel: if that is their aim they could use the existing `concludeAndTransferAll` method.

What seems a more likely scenario to me: upon getting a conclusion proof, Wallet A wants to close the channel and pay out (only) their part of the outcome. Wallet B then needs to `transfer` out the remaining part of the outcome. We should support this scenario and your suggestion with indices ought to work well.

Reply:
>  think what is confusing me is that in both of your strategies, you mention one wallet making calls to fully defund a channel

I'm not sure I follow this. Are you referring to this step?

> 1. Call `concludeAndTransfer`, transferring out as many outcome items as they're willing. 

This step indicates calling a function that does not exist, called `concludeAndTransfer`. It does not indicate fully defunding a channel. It indicates transferring whatever outcome items are specified, which may have the effect of fully defunding the channel.

Reply:
Yes that step makes sense; I was referring to this step 

> 2. While the adjudicator exists, call transfer to iteratively transfer out remaining funds.

Is this the same wallet as in 1.?

Reply:
It can be either wallet! Since each transaction needs to supply the pre-image of the hashed state, the protocol should either decide on a standard sequence of transactions, or the peers should negotiate a sequence of transactions.

Reply:
I think we are on the same page here. I was just (mildly) confused because the way you wrote it, it seemed as if all the transactions came from the same wallet. Of course we would support that, but I think it doesn't quite expose the main benefit?